File: Parallel Programming - Weeks 4-7/Week 5/5 - Thread Cooperation/thread_cooperation.c
/*
 * The C version of ThreadCooperation.java. Same idea: one
 * deposit thread, one withdraw thread, one shared balance. The
 * withdraw thread waits when the balance is too small.
 *
 * Where Java used
 *     synchronized + lock.wait() + lock.notifyAll()
 * on a single lock object, C splits the job between two things:
 *     pthread_mutex_t mutex    — the lock
 *     pthread_cond_t  cv       — the waiting room
 *
 * They are two separate objects, but they always travel
 * together. Every pthread_cond_wait takes the matching mutex as
 * its second argument.
 *
 * Compile: gcc thread_cooperation.c -o thread_cooperation -lpthread
 * Run:     ./thread_cooperation
 * Stop:    Ctrl+C (the loops never end on their own)
 */

#include <stdio.h>
#include <stdlib.h>    /* rand, srand */
#include <time.h>      /* time */
#include <unistd.h>    /* sleep */
#include <pthread.h>

int balance = 0;
pthread_mutex_t mutex;
pthread_cond_t  cv;

void* deposit_loop(void* arg) {
    while (1) {
        pthread_mutex_lock(&mutex);

        int amount = rand() % 10 + 1;
        balance += amount;
        printf("Add %d. Balance now is %d\n", amount, balance);

        /* Ring the bell. Same role as lock.notifyAll() in Java. */
        pthread_cond_broadcast(&cv);

        pthread_mutex_unlock(&mutex);
        sleep(1);
    }
    return NULL;
}

void* withdraw_loop(void* arg) {
    while (1) {
        pthread_mutex_lock(&mutex);

        int amount = rand() % 20 + 1;

        /* while, not if. Same reason as in the Java version:
           after waking up, the condition may still be false. */
        while (balance < amount) {
            printf("\t\t\tCannot withdraw %d, waiting...\n", amount);

            /* Same role as lock.wait() in Java. Releases the
               mutex, sleeps, and reacquires the mutex on wakeup. */
            pthread_cond_wait(&cv, &mutex);
        }

        balance -= amount;
        printf("\t\t\tWithdrew %d. Balance now is %d\n", amount, balance);

        pthread_mutex_unlock(&mutex);
    }
    return NULL;
}

int main() {
    pthread_t d, w;

    pthread_mutex_init(&mutex, NULL);
    pthread_cond_init(&cv, NULL);
    srand(time(NULL));

    pthread_create(&d, NULL, deposit_loop, NULL);
    pthread_create(&w, NULL, withdraw_loop, NULL);

    /* These never actually return because the workers loop
       forever. They're here for completeness; in real code you
       would join threads that have a finite lifetime. */
    pthread_join(d, NULL);
    pthread_join(w, NULL);

    pthread_mutex_destroy(&mutex);
    pthread_cond_destroy(&cv);
    return 0;
}

File: Parallel Programming - Weeks 4-7/Week 5/5 - Thread Cooperation/ThreadCooperation.java
/*
 * Two threads sharing one bank account.
 *
 * The deposit thread adds a random 1..10 once per second. The
 * withdraw thread tries to take out a random 1..20 and, when the
 * balance is too small, waits on the lock object until a deposit
 * wakes it up.
 *
 * Both threads loop forever. Let the program run for a few
 * seconds, watch the two clerks trade places, then stop it with
 * Ctrl+C.
 *
 * The C version with pthread_cond_t is in thread_cooperation.c
 * next to this file. It has the same structure; only the API
 * names and the manual unlock are different.
 */
import java.util.Random;

public class ThreadCooperation {

    static int balance = 0;
    static final Object lock = new Object();
    static final Random rand = new Random();

    public static void main(String[] args) {
        // Two infinite workers. We don't bother joining because
        // neither of them ever finishes.
        new Thread(ThreadCooperation::depositLoop).start();
        new Thread(ThreadCooperation::withdrawLoop).start();
    }

    static void depositLoop() {
        while (true) {
            synchronized (lock) {
                int amount = rand.nextInt(10) + 1;
                balance += amount;
                System.out.println("Add " + amount + ". Balance now is " + balance);

                // Ring the bell. Wakes up anyone currently in wait()
                // on this same lock object.
                lock.notifyAll();
            }
            try { Thread.sleep(1000); } catch (InterruptedException e) {}
        }
    }

    static void withdrawLoop() {
        while (true) {
            synchronized (lock) {
                int amount = rand.nextInt(20) + 1;

                // while, not if. See the README for the reason.
                while (balance < amount) {
                    System.out.println("\t\t\tCannot withdraw " + amount + ", waiting...");
                    try {
                        lock.wait();
                    } catch (InterruptedException e) {
                        return;
                    }
                }

                balance -= amount;
                System.out.println("\t\t\tWithdrew " + amount + ". Balance now is " + balance);
            }
        }
    }
}
